File: 15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */

// @lc code=start
#include <vector>
#include <algorithm>
using std::sort;
using std::vector;

class Solution {
public:
    vector<vector<int> > threeSum(vector<int>& nums) {
        vector<vector<int> > res;
        sort(nums.begin(), nums.end());
        int i = 0;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i-1] == nums[i]) continue;
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right) {
                if (!(nums[i] + nums[left] + nums[right])) {
                    res.push_back({nums[i], nums[left], nums[right]});
                    while ((left < right) && (nums[left+1] == nums[left])) ++left;
                    while ((right > 1) && (nums[right-1] == nums[right])) --right;
                    ++left;
                    --right;
                } else if (nums[i] + nums[left] + nums[right] > 0) 
                    --right;
                else
                    ++left;
            }
        }
        return res;
    }
};
// @lc code=end

File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#
# https://leetcode.com/problems/3sum/description/
#
# algorithms
# Medium (25.64%)
# Likes:    5542
# Dislikes: 670
# Total Accepted:    770.9K
# Total Submissions: 3M
# Testcase Example:  '[-1,0,1,2,-1,-4]'
#
# Given an array nums of n integers, are there elements a, b, c in nums such
# that a + b + c = 0? Find all unique triplets in the array which gives the sum
# of zero.
#
# Note:
#
# The solution set must not contain duplicate triplets.
#
# Example:
#
#
# Given array nums = [-1, 0, 1, 2, -1, -4],
#
# A solution set is:
# [
# ⁠ [-1, 0, 1],
# ⁠ [-1, -1, 2]
# ]
#
#
from typing import List


# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ret = []
        nums.sort()
        if not nums:
            return ret
        a = None
        for i in range(len(nums) - 2):
            if nums[i] == a:
                continue
            a = nums[i]
            if a > 0:
                break
            j = i + 1
            k = len(nums) - 1
            while j < k:
                b = nums[j]
                c = nums[k]
                if a + b + c < 0:
                    j += 1
                elif a + b + c > 0:
                    k -= 1
                else:
                    ret.append([nums[i], nums[j], nums[k]])
                    j += 1
                    while j < k and nums[j] == b:
                        j += 1
                    k -= 1
                    while j < k and nums[k] == c:
                        k -= 1
        return ret


# @lc code=end
